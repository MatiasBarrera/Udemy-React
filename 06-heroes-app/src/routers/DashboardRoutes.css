.container {
  padding: 1rem 5rem;
  height: 90vh;
}
