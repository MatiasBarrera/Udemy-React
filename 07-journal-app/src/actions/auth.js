import {
  getAuth,
  signInWithPopup,
  createUserWithEmailAndPassword,
  updateProfile,
  signInWithEmailAndPassword,
  signOut
} from '@firebase/auth'
import Swal from 'sweetalert2'
import { googleAuthProvider } from '../firebase/firebase-config'
import { types } from '../types/types'
import { FinishLoading, StartLoading } from './ui'

export const startLoginEmailPassword = (email, password) => {
  return dispatch => {
    const auth = getAuth()

    dispatch(StartLoading())
    signInWithEmailAndPassword(auth, email, password)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName))
        dispatch(FinishLoading())
      })
      .catch(e => {
        Swal.fire({
          icon: 'error',
          title: 'Error',
          text: 'There is no user record corresponding to this identifier. The user may have been deleted.'
        })
        dispatch(FinishLoading())
      })
  }
}

export const startRegisterWithEmailPasswordName = (name, email, password) => {
  return dispatch => {
    const auth = getAuth()
    createUserWithEmailAndPassword(auth, email, password)
      .then(async ({ user }) => {
        await updateProfile(user, { displayName: name })

        dispatch(login(user.uid, user.displayName))
      })
      .catch(e => {
        console.log(e)
      })
  }
}

export const startGoogleLogin = () => {
  return dispatch => {
    const auth = getAuth()
    signInWithPopup(auth, googleAuthProvider).then(({ user }) => {
      dispatch(login(user.uid, user.displayName))
    })
  }
}

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    name: displayName
  }
})

export const startLogout = () => {
  return async dispatch => {
    const auth = getAuth()
    await signOut(auth)
    dispatch(logout())
  }
}

export const logout = () => ({
  type: types.logout
})
