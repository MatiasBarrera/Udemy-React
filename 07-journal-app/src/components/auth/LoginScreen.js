import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Link } from 'react-router-dom'
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth'
import googleLogo from '../../assets/google.logo.svg'
import { useForm } from '../../hooks/useForm'

export const LoginScreen = () => {
  const dispatch = useDispatch()
  const { loading } = useSelector(state => state.ui)

  const [formValues, handleInputChange] = useForm({
    email: 'matias@gmail.com',
    password: '123456'
  })

  const handleLogin = e => {
    e.preventDefault()
    dispatch(startLoginEmailPassword(email, password))
  }

  const handleLoginGoogle = () => {
    dispatch(startGoogleLogin())
  }

  const { email, password } = formValues
  return (
    <>
      <h3 className="login__title">Login</h3>
      <form className="auth__form" onSubmit={handleLogin}>
        <input
          type="text"
          placeholder="email"
          name="email"
          autoComplete="off"
          value={email}
          onChange={handleInputChange}
        />
        <input
          type="password"
          placeholder="password"
          name="password"
          autoComplete="off"
          value={password}
          onChange={handleInputChange}
        />
        <button className="pointer" type="submit" disabled={loading}>
          Ingresar
        </button>
        <hr />

        <div className="login__social-network">
          <p>Login with social network</p>
          <div
            className="google-btn pointer mt-1 mb-1"
            onClick={handleLoginGoogle}>
            <div className="google-icon-wrapper">
              <img
                className="google-icon"
                src={googleLogo}
                alt="google button"
              />
            </div>
            <p className="btn-text">
              <b>Sign in with google</b>
            </p>
          </div>
        </div>

        <Link to="/auth/register">Create new account</Link>
      </form>
    </>
  )
}
