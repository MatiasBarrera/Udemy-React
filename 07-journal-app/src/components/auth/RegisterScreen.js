import React from 'react'
import { Link } from 'react-router-dom'
import { useForm } from '../../hooks/useForm'
import isEmail from 'validator/lib/isEmail'
import { useDispatch, useSelector } from 'react-redux'
import { removeError, setError } from '../../actions/ui'
import { startRegisterWithEmailPasswordName } from '../../actions/auth'

export const RegisterScreen = () => {
  const dispatch = useDispatch()
  const { msgError } = useSelector(state => state.ui)

  const [formValues, handleInputChange] = useForm({
    name: 'matias',
    email: 'matias@gmail.com',
    password: '123456',
    password2: '123456'
  })

  const { name, email, password, password2 } = formValues

  const handleRegister = e => {
    e.preventDefault()

    if (isFormValid()) {
      dispatch(startRegisterWithEmailPasswordName(name, email, password))
    }
  }

  const isFormValid = () => {
    if (name.trim().length === 0) {
      dispatch(setError('Name is required'))
      return false
    }
    if (email.trim().length === 0 && !isEmail(email)) {
      dispatch(setError('Email is not valid'))
      return false
    }
    if (password !== password2 || password.length < 5) {
      dispatch(
        setError(
          'Password should be at least 6 characters and match each other '
        )
      )
      return false
    }
    dispatch(removeError())
    return true
  }

  return (
    <>
      <h3 className="login__title">Register</h3>
      <form className="auth__form" onSubmit={handleRegister}>
        {msgError && <div className="auth__alert-error">{msgError}</div>}
        <input
          type="text"
          placeholder="name"
          name="name"
          autoComplete="off"
          value={name}
          onChange={handleInputChange}
        />
        <input
          type="text"
          placeholder="email"
          name="email"
          autoComplete="off"
          value={email}
          onChange={handleInputChange}
        />
        <input
          type="password"
          placeholder="password"
          name="password"
          autoComplete="off"
          value={password}
          onChange={handleInputChange}
        />
        <input
          type="password"
          placeholder="confirm"
          name="password2"
          autoComplete="off"
          value={password2}
          onChange={handleInputChange}
        />
        <button className="pointer" type="submit">
          Register
        </button>
        <hr />

        <Link to="/auth/login">Already registered?</Link>
      </form>
    </>
  )
}
