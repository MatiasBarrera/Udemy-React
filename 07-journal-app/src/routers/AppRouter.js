import { getAuth, onAuthStateChanged } from '@firebase/auth'
import React, { useEffect, useState } from 'react'
import { useDispatch } from 'react-redux'
import { login } from '../actions/auth'

import {
  BrowserRouter as Router,
  Switch,
  // Route,
  Redirect
} from 'react-router-dom'
import { JournalScreen } from '../components/journal/JournalScreen'
import { AuthRouter } from './AuthRouter'
import { PublicRoute } from './PublicRoute'
import { PrivateRoute } from './PrivateRoute'

export const AppRouter = () => {
  const dispatch = useDispatch()

  const [checking, setChecking] = useState(true)

  const [isLoggedIn, setIsLoggedIn] = useState(false)

  useEffect(() => {
    const auth = getAuth()
    onAuthStateChanged(auth, user => {
      if (user?.uid) {
        dispatch(login(user.uid, user.displayName))
        setIsLoggedIn(true)
      } else {
        setIsLoggedIn(false)
      }
      setChecking(false)
    })
  }, [dispatch, setChecking, setIsLoggedIn, isLoggedIn])

  if (checking) {
    return <h1>Cargando...</h1>
  }

  return (
    <Router>
      <div>
        <Switch>
          <PublicRoute
            path="/auth"
            component={AuthRouter}
            isLoggedIn={isLoggedIn}
          />
          <PrivateRoute
            exact
            isLoggedIn={isLoggedIn}
            path="/"
            component={JournalScreen}
          />

          <Redirect to="/auth/login" />
        </Switch>
      </div>
    </Router>
  )
}
