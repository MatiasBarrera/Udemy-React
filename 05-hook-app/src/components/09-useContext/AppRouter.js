import React from 'react'
import { BrowserRouter as Router, Switch, Route } from 'react-router-dom'

import { Navbar } from './Navbar'

import { AboutScreen } from './AboutScreen'
import { LoginScreen } from './LoginScreen'
import { HomeScreen } from './HomeScreen'
import { NotFoundScreen } from './NotFoundScreen'

export const AppRouter = () => {
  return (
    <Router>
      <>
        <Navbar />
        <div className="container">
          <Switch>
            <Route exact path="/" component={HomeScreen} />
            <Route exact path="/about" component={AboutScreen} />
            <Route exact path="/login" component={LoginScreen} />

            <Route exact component={NotFoundScreen} />
          </Switch>
        </div>
      </>
    </Router>
  )
}
