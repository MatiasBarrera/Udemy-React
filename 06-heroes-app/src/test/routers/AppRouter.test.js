import React from 'react'
import { mount } from 'enzyme'
import { AppRouter } from '../../routers/AppRouter'
import { AuthContext } from '../../auth/AuthContext'

describe('Pruebas en AppRouter', () => {
  const contextValue = {
    dispatch: jest.fn(),
    user: {
      logged: false
    }
  }

  test('Debe de mostrar login si no esta autenticado', () => {
    const wrapper = mount(
      <AuthContext.Provider value={contextValue}>
        <AppRouter />
      </AuthContext.Provider>
    )
    expect(wrapper).toMatchSnapshot()
  })

  test('Debe de mostrar el componenete de Marvel si esta autenticado', () => {
    const contextValue = {
      dispatch: jest.fn(),
      user: {
        logged: true,
        name: 'Matias'
      }
    }

    const wrapper = mount(
      <AuthContext.Provider value={contextValue}>
        <AppRouter />
      </AuthContext.Provider>
    )

    expect(wrapper.find('nav').exists()).toBeTruthy()
  })
})
