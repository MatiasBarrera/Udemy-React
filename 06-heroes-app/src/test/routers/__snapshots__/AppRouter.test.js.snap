// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en AppRouter Debe de mostrar login si no esta autenticado 1`] = `
<div>
  <div
    className="login-container"
  >
    <div
      className="login-card-container"
    >
      <div
        className="login-card-header"
      >
        <h1>
          Welcome
        </h1>
      </div>
      <div
        className="login-card-body"
      >
        <button
          onClick={[Function]}
        >
          Login
        </button>
      </div>
    </div>
  </div>
</div>
`;
